### PR TITLE
mailpit: implement `passthru.updateScript`

### PR DESCRIPTION
**Pull request by RafaelKr**
## Description of changes

Supersedes #324026 with the following improvements:
- Remove the leading `v` from the version number
- Correctly indent `versions.nix`
- Don't update the version with this PR yet, only implement the update script. Let's wait for @r-ryantm bot to create the PR to ensure it's also working in the real context. (Of course I tested it locally with `nix-shell maintainers/scripts/update.nix --argstr package mailpit --argstr commit true`)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/mail/mailpit/default.nix [hunk 1]
```diff
@@ -56,11 +64,15 @@ buildGoModule {
   pname = "mailpit";
   inherit src version;
 
-  vendorHash = "sha256-e2mlOwGDU5NlKZSstHMdTidSfhNeeY6cBgtW+W9nwV8=";
+  vendorHash = versions.vendorHash;
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/mailpit/default.nix:68]:
```suggestion
```

File: pkgs/servers/mail/mailpit/default.nix [hunk 2]
```diff
@@ -1,24 +1,27 @@
...
+  versions = import ./versions.nix;
+
+  version = versions.version;
 
   src = fetchFromGitHub {
     owner = "axllent";
     repo = "mailpit";
     rev = "v${version}";
-    hash = "sha256-5QEn4sEZgtoFrVonZsAtvzhEkxYGDEWwhJOxqwWQJTk=";
+    hash = versions.hash;
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/mailpit/default.nix:24]:
```suggestion
    inherit (source) hash;
```

File: pkgs/servers/mail/mailpit/default.nix [hunk 3]
```diff
@@ -30,7 +33,7 @@ let
 
     npmDeps = fetchNpmDeps {
       inherit src;
-      hash = "sha256-5F68ia2V8mw4iPAjSoz0b8z1lplWtAg98BgDXYOmMKs=";
+      hash = versions.npmDepsHash;
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/mailpit/default.nix:36]:
```suggestion
      inherit (source) npmDepsHash;
```

File: pkgs/servers/mail/mailpit/versions.nix
Comment (SuperSandro2000) [inline pkgs/servers/mail/mailpit/versions.nix:1]: versions implies that there are multiple versions which is not the case

File: pkgs/servers/mail/mailpit/update.sh [hunk 4]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p nix coreutils curl jq nix-prefetch-git prefetch-npm-deps
+
+set -x -euo pipefail
+
+OWNER=axllent
+REPO=mailpit
+
+TARGET_TAG="$(curl https://api.github.com/repos/$OWNER/$REPO/releases/latest | jq -r '.tag_name')"
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/mailpit/update.sh:9]: This should optionally take a GITHUB_TOKEN

```suggestion
TARGET_TAG="$(curl -L -s ${GITHUB_TOKEN:+-u ":${GITHUB_TOKEN}"} https://api.github.com/repos/axllent/mailpit/releases/latest | jq -r '.tag_name')"
```

File: pkgs/servers/mail/mailpit/update.sh [hunk 5]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p nix coreutils curl jq nix-prefetch-git prefetch-npm-deps
+
+set -x -euo pipefail
+
+OWNER=axllent
+REPO=mailpit
+
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/mailpit/update.sh:8]:
```suggestion
```

File: pkgs/servers/mail/mailpit/update.sh [hunk 6]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p nix coreutils curl jq nix-prefetch-git prefetch-npm-deps
+
+set -x -euo pipefail
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/mailpit/update.sh:4]:
```suggestion
set -euo pipefail
```

File: pkgs/servers/mail/mailpit/update.sh [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+TMP="$(mktemp -d)"
+trap 'rm -rf "$TMP"' exit
+
+NIXPKGS_MAILPIT_PATH=$(cd $(dirname ${BASH_SOURCE[0]}); pwd -P)/
+VERSIONS_NIX="$NIXPKGS_MAILPIT_PATH/versions.nix"
+
+PREFETCH_JSON=$TMP/prefetch.json
+nix-prefetch-git --rev "$TARGET_TAG" --url https://github.com/$OWNER/$REPO > "$PREFETCH_JSON"
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/mailpit/update.sh:34]:
```suggestion
nix-prefetch-git --rev "$TARGET_TAG" --url https://github.com/axllent/mailpit > "$PREFETCH_JSON"
```

**Comment by RafaelKr** [comment]
@SuperSandro2000 Thanks for all your suggestions!

The one thing I don't get is the change from `source.<var>` to `inherit (source) <var>`. Is there any performance benefit like lazy loading or is it convention or what's the reasoning?

File: pkgs/servers/mail/mailpit/update.sh (hunk 5 above)
Comment (RafaelKr) [reply pkgs/servers/mail/mailpit/update.sh:8]: I'd like to keep that as it's used multiple times (L9 and L34). So if the Repo would be moved in the future it only needs to be changed once and no occurrence can be missed.

File: pkgs/servers/mail/mailpit/default.nix (hunk 3 above)
Comment (RafaelKr) [reply pkgs/servers/mail/mailpit/default.nix:36]: I think this wouldn't work, as we need the attribute key `hash`. With your change the key would be `npmDepsHash`.

File: pkgs/servers/mail/mailpit/default.nix (hunk 2 above)
Comment (RafaelKr) [reply pkgs/servers/mail/mailpit/default.nix:24]: I don't really get this suggestion. Is there any performance benefit like lazy loading or is it convention or what's the reasoning?

I would rather keep using `hash = source.hash;`

**Comment by RafaelKr** [comment]
@SuperSandro2000 I just pushed some of your suggested changes, thanks again!

I resolved all conversations which are done and kept the others open for further discussion :)

**Comment by RafaelKr** [comment]
Hey @SuperSandro2000, sorry for pinging, but I would be happy hearing from you, thanks in advance! :)

**Comment by RafaelKr** [comment]
There's a new mailpit release fixing a stored XSS vulnerability: https://github.com/axllent/mailpit/releases/tag/v1.19.3

I'd like to get mailpit updated in nixpkgs. Should I create a separate PR for that version or can we get this PR merged and @r-ryantm will do it for us?

**Comment by SuperSandro2000** [comment]
You can do a PR for an update of a package any time. No need to wait for the bot if things are more complicated.

**Pull request merged**